block-node/messaging/src/test/java/org/hiero/block/server/messaging/BlockMessagingServiceBlockNotificationTest.java

```diff
@@ -163,66 +144,35 @@ void testBlockNotificationBackpressure() throws InterruptedException {
-            if (TEST_DATA_COUNT == receivedCounter.incrementAndGet()) {
-                // call the latch countdown when we reach the expected count
-                latch.countDown();
-            }
-            // slow down the handler, if sending is not finished
-            try {
-                holdBackSlowHandler.acquire();
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            }
-        };
+        BlockNotificationHandler slowHandler =
+                new LatchCountdownHandler(TEST_DATA_COUNT, counters, latch, 0, holdBackSlowHandler);
         // Create MessagingService to test and register the handlers
         BlockMessagingFacility messagingService = new BlockMessagingFacilityImpl();
-        messagingService.init(BLOCK_NODE_CONTEXT, null);
+        messagingService.init(TestConfig.BLOCK_NODE_CONTEXT, null);
         messagingService.registerBlockNotificationHandler(slowHandler, false, null);
         // start the messaging service
         messagingService.start();
         // send TEST_DATA_COUNT block notifications in a background thread
-        Thread senderThread = new Thread(() -> {
-            for (int i = 0; i < TEST_DATA_COUNT; i++) {
-                messagingService.sendBlockNotification(new BlockNotification(i, BLOCK_PERSISTED, null));
-                sentCounter.incrementAndGet();
-                // release the slow handler every other time so it is slowed down by 50%
-                if (i % 4 == 0) {
-                    holdBackSlowHandler.release(1);
-                }
-            }
-        });
+        Thread senderThread =
+                new SendingThread(messagingService, sentCounter, counters, holdBackSlowHandler, 0, 0.75D, 3);
         senderThread.start();
-        // wait for the sender thread to get stuck in back pressure
-        while (senderThread.getState() == State.RUNNABLE) {
-            try {
-                //noinspection BusyWait
-                Thread.sleep(2);
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            }
-        }
-        // the send thread should get blocked in back pressure
-        final var stateAfter = senderThread.getState();
-        assertNotEquals(
-                State.RUNNABLE,
-                stateAfter,
-                "Sender thread should be blocked in back pressure, but is in state: " + stateAfter);
-        // only 250 ish notifications should have been sent to the messaging service
-        assertTrue(sentCounter.get() > 100, "sentCounter = " + sentCounter.get() + " is not > 100");
-        assertTrue(
-                sentCounter.get() < TEST_DATA_COUNT,
-                "sentCounter = " + sentCounter.get() + " is not < " + TEST_DATA_COUNT);
+        final State stateAfter = waitForSenderBlockedOrComplete(senderThread, 1, TimeUnit.SECONDS);
+        assertNotEquals(State.RUNNABLE, stateAfter, "Sender thread should not be runnable.");
```

Comment:
Q: would is better to assert for the desired state?
Since you called `waitForSenderBlockedOrComplete(...)` could this be simplified to?
```suggestion
        assertEquals(State.BLOCKED, stateAfter, "Sender thread should be blocked.");
```

Reply:
It could be any of 3 different states, so the assert becomes quite problematic that way.
The simplest "correct" assertion is that it is neither running nor terminated.

Reply:
Fun times 😅